README.md
# BF Visual Interpreter [1.0.0]

## Purpose

This Python Application is a visual interpreter for the [Brainfuck](https://en.wikipedia.org/wiki/Brainfuck) programming language. It is intended as a way to solve pointless problems in a useless language in a way that will not make your head explode. Because if you gamify something horrible and useless, you get something tolerable and useless.

### Notes

- The ultimate goal is to set *challenges* by providing and initial memory set and a expected results. 
- The user(see: [masochist](https://en.wikipedia.org/wiki/Masochism_(disambiguation))) then provides the program to transition from the initial to target states when the program halts.
- At this point you need to make a goal on your own and provide your own critique

## Instructions

This interpreter takes its initial memory layout and instruction set from command line values, and runs them through what is hopefully a reasonable-ish user interface. This section defines available inputs and controls.

### Command Line Arguments

| Short | Long | Effect |
| --- | --- | --- |
| -v | --verbose | Sets logging to *DEBUG* for in depth execution information. |
| -s | --src-file | **REQUIRED** Specifies a file containing a BF program source file. |
| -e | --env-file | Specifies a file defining the environment the BF program will execute within. If not specified will default to 8 cells with a maximum value of 16 |

### Source Files

Source files must be in restricted BF format, containing only the following characters. If you wanted comments and easy readability, I have to question your judgement in learning this particular lenguage. 

In seriousness, this is a concious choice to facilitate future additions for the [Brain Function](https://github.com/ryanfox/brainfunction) extensions.

`+-><[].,`

#### Notes

- No whitespaces, newlines, or other characters are accepted in this version.
- The character ',' will prompt for user input in the visualizer
- The character '.' will print the value of the current cell to the console regardless of verbosity. 
    - There is a planned update to display this to the screen in a popup for full visual support

### Environment Configurations

The environment configurations are provided by JSON files. All of the tags must be included. Several samples are shown below for reference.

An environment with:
- 4 cells
- Max value per cell of 24
- All cell values set to 3 at the start

```json
{
    "version": "1",
    "memory": {
        "cell_count": 4,
        "cell_max_value": 24,
        "cell_default_value": 3,
        "cell_initial_values": []
    }
}
```

An environment with:
- 8 cells
- Max value per cell of 16
- All unspecified cell values set to 0 at the start
- The first six values are set to the sequence 1,2,3,4,5,6

```json
{
    "version": "1",
    "memory": {
        "cell_count": 8,
        "cell_max_value": 16,
        "cell_default_value": 0,
        "cell_initial_values": [1,2,3,4,5,6]
    }
}
```

### Sample Executions

Specifying all values with detailed outputs.

```ps1
VisInt> python.exe .\main.py -s "samples/prg_1.bf" -e "samples/env_1.json" --verbose
VisInt> python.exe .\main.py --src-file "samples/prg_1.bf" --env-file "samples/env_1.json" --verbose
```

Executing *samples/prg_1.bf* in the default environment.

```ps1
VisInt> python.exe .\main.py -s "samples/prg_1.bf"
```

### Controls

In the interpreter itself, there are a few controls you have during program execution.

#### Execution Speed

| Control | Effect | Detail |
| --- | --- | --- |
| Spacebar | Toggle Pause | The interpreter starts *Paused*. You must press spacebar to begin running the Brainfuck program |
| Tab | Increment Execution Speed | Pressing *TAB* will double the instructions per second speed. It will go from 1HZ to a maximum of 32 HZ, before rolling over back to 1 HZ. |

#### Entering values on Prompt

Entering cell values via the `,` command is supported by an on-screen prompt. When the interpreter executes this command the execution will pause until a number in the allowed value range (0 -> maximum value, default 16) is entered. Once entered the current cell will be assigned that value provided.

Entering an illegal value (outside of the allowed range) will result in a program execution error.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.0.1] - 2022-09-28
### Added
- Hardcoded rendering configuration
    - Screen scrolls fluidly to keep program counter in center
    - Each cell is represented by a red base with growing white tower
- Loading bf programs from CLI
- Loading bf memory environments from CLI
- Unit tests for BF Interpreting engine
- Screen elements to display current run configuration
- Controls to change the execution rate for powers of 2 in the range 2^0 -> 2^5
- Pause to temporarily halt execution
- Initial README file

## [0.0.2] - 2022-10-22
### Added
- Input prompt support for ',' command

### Changed

- Refactored rendering to classes outside of main. Now main is responsible for event handling and state controls only.


tests/test_bf.py
import pytest

from src.bf import BFCommand, BFInitError, BFInterpreter, BFRuntimeError, ProgramState


# Groups tests related to initialization of a BF Interpreter
class TestBFInit:

    def test_initialization(self):
        interpreter = BFInterpreter(5)
        interpreter.setTape("+++")

        assert interpreter.ptr == 0
        assert interpreter.pc == 0
        assert interpreter.state == ProgramState.Ready
        for cell in interpreter.memory:
            assert cell == 0

    def test_setTape(self):
        interpreter = BFInterpreter(1)
        interpreter.setTape("+-<>[].,")
        print(interpreter.tapeString())

        assert len(interpreter.tape) == len("+-<>[].,")
        assert interpreter.tape[0] == BFCommand.Increment
        assert interpreter.tape[1] == BFCommand.Decrement
        assert interpreter.tape[2] == BFCommand.CellPtrLeft
        assert interpreter.tape[3] == BFCommand.CellPtrRight
        assert interpreter.tape[4] == BFCommand.StartWhile
        assert interpreter.tape[5] == BFCommand.EndWhile
        assert interpreter.tape[6] == BFCommand.PrintByte
        assert interpreter.tape[7] == BFCommand.ReadByte

    def test_init_memory_size_error(self):

        with pytest.raises(BFInitError):
            BFInterpreter(0)

    def test_init_max_value_error(self):

        with pytest.raises(BFInitError):
            BFInterpreter(1, 0)

    def test_init_empty_tape_error(self):

        with pytest.raises(BFInitError):
            BFInterpreter(1).setTape("")


# Groups tests related to execution of a BF Interpreter
class TestBFRuntime:

    def test_step_to_halt(self):
        interpreter = BFInterpreter(1)
        interpreter.setTape("+++")

        assert interpreter.pc == 0
        assert interpreter.state == ProgramState.Ready

        interpreter.step()

        assert interpreter.pc == 1
        assert interpreter.state == ProgramState.Running

        interpreter.step()

        assert interpreter.pc == 2
        assert interpreter.state == ProgramState.Running

        interpreter.step()

        assert interpreter.pc == 3
        assert interpreter.state == ProgramState.Halted

    def test_pointer_commands(self):
        interpreter = BFInterpreter(2)
        interpreter.setTape("><")

        assert interpreter.ptr == 0
        interpreter.step()
        assert interpreter.ptr == 1
        interpreter.step()
        assert interpreter.ptr == 0

    def test_increment_decrement_commands(self):
        interpreter = BFInterpreter(1)
        interpreter.setTape("+-")

        assert interpreter.memory[interpreter.ptr] == 0
        interpreter.step()
        assert interpreter.memory[interpreter.ptr] == 1
        interpreter.step()
        assert interpreter.memory[interpreter.ptr] == 0

    def test_simple_loop_commands(self):
        interpreter = BFInterpreter(1)
        interpreter.setTape("+[-]")

        assert interpreter.pc == 0

        interpreter.step()
        assert interpreter.pc == 1

        interpreter.step()
        assert interpreter.pc == 2

        interpreter.step()
        assert interpreter.pc == 3

        interpreter.step()
        assert interpreter.pc == 1

        interpreter.step()
        assert interpreter.pc == 4
        assert interpreter.state == ProgramState.Halted

    def test_nested_loop_commands(self):
        interpreter = BFInterpreter(1)
        interpreter.setTape("+[[-]]")

        assert interpreter.pc == 0

        interpreter.step()
        assert interpreter.pc == 1

        interpreter.step()
        assert interpreter.pc == 2

        interpreter.step()
        assert interpreter.pc == 3

        interpreter.step()
        assert interpreter.pc == 4

        interpreter.step()
        assert interpreter.pc == 2

        interpreter.step()
        assert interpreter.pc == 5

        interpreter.step()
        assert interpreter.pc == 1

        interpreter.step()
        assert interpreter.pc == 6
        assert interpreter.state == ProgramState.Halted

    def test_runtime_underflow_error(self):
        interpreter = BFInterpreter(1)
        interpreter.setTape("-")

        with pytest.raises(BFRuntimeError):
            interpreter.step()

    def test_runtime_overflow_error(self):
        interpreter = BFInterpreter(1, 1)
        interpreter.setTape("++")

        with pytest.raises(BFRuntimeError):
            interpreter.step()
            interpreter.step()

    def test_runtime_ptr_less_than_zero_error(self):
        interpreter = BFInterpreter(1)
        interpreter.setTape("<")

        with pytest.raises(BFRuntimeError):
            interpreter.step()

    def test_runtime_ptr_out_of_bounds_error(self):
        interpreter = BFInterpreter(1)
        interpreter.setTape(">")

        with pytest.raises(BFRuntimeError):
            interpreter.step()

    def test_runtime_step_after_halt_error(self):
        interpreter = BFInterpreter(1)
        interpreter.setTape("+")

        interpreter.step()

        with pytest.raises(BFRuntimeError):
            interpreter.step()

    def test_runtime_step_after_error_error(self):
        interpreter = BFInterpreter(1)
        interpreter.setTape("+")

        interpreter.step()

        # Initial Error
        with pytest.raises(BFRuntimeError):
            interpreter.step()

        # Error caused by trying to step again
        with pytest.raises(BFRuntimeError):
            interpreter.step()

    def test_setMemory_defaults_only(self):
        interpreter = BFInterpreter(8, 16)
        interpreter.setMemory(values=[], default=0)

        assert len(interpreter.memory) == 8
        for cell in interpreter.memory:
            assert cell == 0

    def test_setMemory_initialized_and_default(self):
        interpreter = BFInterpreter(8, 16)
        interpreter.setMemory(values=[1, 2, 3, 4], default=8)

        assert len(interpreter.memory) == 8
        assert interpreter.memory[0] == 1
        assert interpreter.memory[1] == 2
        assert interpreter.memory[2] == 3
        assert interpreter.memory[3] == 4

        for i in range(4, len(interpreter.memory)):
            assert interpreter.memory[i] == 8

    def test_setMemory_fully_initialized(self):
        interpreter = BFInterpreter(8, 16)
        interpreter.setMemory(values=[1, 2, 3, 4, 5, 6, 7, 8], default=0)

        assert len(interpreter.memory) == 8
        assert interpreter.memory[0] == 1
        assert interpreter.memory[1] == 2
        assert interpreter.memory[2] == 3
        assert interpreter.memory[3] == 4
        assert interpreter.memory[4] == 5
        assert interpreter.memory[5] == 6
        assert interpreter.memory[6] == 7
        assert interpreter.memory[7] == 8

    def test_setMemory_negative_default_error(self):
        interpreter = BFInterpreter(8, 16)
        with pytest.raises(BFInitError):
            interpreter.setMemory(values=[], default=-1)

    def test_setMemory_large_default_error(self):
        interpreter = BFInterpreter(8, 16)
        with pytest.raises(BFInitError):
            interpreter.setMemory(values=[], default=17)

    def test_setMemory_long_values_error(self):
        interpreter = BFInterpreter(8, 16)
        with pytest.raises(BFInitError):
            interpreter.setMemory(values=[1, 2, 3, 4, 5, 6, 7, 8, 9], default=0)

    def test_setMemory_negative_values_error(self):
        interpreter = BFInterpreter(8, 16)
        with pytest.raises(BFInitError):
            interpreter.setMemory(values=[1, -1], default=0)

    def test_setMemory_large_values_error(self):
        interpreter = BFInterpreter(8, 16)
        with pytest.raises(BFInitError):
            interpreter.setMemory(values=[1, 17], default=0)

    def test_readByte_success_readLastStep(self):
        interpreter = BFInterpreter(8, 16)
        interpreter.setMemory([], 0)
        interpreter.setTape(",")

        interpreter.step()
        assert interpreter.state == ProgramState.WaitingForInput
        assert interpreter.pc == 0
        assert interpreter.memory[interpreter.ptr] == 0

        interpreter.readByte(5)
        interpreter.state == ProgramState.Halted
        assert interpreter.pc == 1
        assert interpreter.memory[interpreter.ptr] == 5

    def test_readByte_success_stepAfterInput(self):
        interpreter = BFInterpreter(8, 16)
        interpreter.setMemory([], 0)
        interpreter.setTape(",")

        interpreter.step()
        assert interpreter.state == ProgramState.WaitingForInput
        assert interpreter.pc == 0
        assert interpreter.memory[interpreter.ptr] == 0

        interpreter.readByte(5)
        interpreter.state == ProgramState.Running
        assert interpreter.pc == 1
        assert interpreter.memory[interpreter.ptr] == 5

    def test_readByte_error_readyStatus(self):
        interpreter = BFInterpreter(8, 16)
        interpreter.setMemory([], 0)
        interpreter.setTape("++")

        assert interpreter.state == ProgramState.Ready
        with pytest.raises(BFRuntimeError):
            interpreter.readByte(0)

    def test_readByte_error_runningStatus(self):
        interpreter = BFInterpreter(8, 16)
        interpreter.setMemory([], 0)
        interpreter.setTape("++")

        interpreter.step()
        assert interpreter.state == ProgramState.Running
        with pytest.raises(BFRuntimeError):
            interpreter.readByte(0)

    def test_readByte_error_haltedStatus(self):
        interpreter = BFInterpreter(8, 16)
        interpreter.setMemory([], 0)
        interpreter.setTape("+")

        interpreter.step()
        assert interpreter.state == ProgramState.Halted
        with pytest.raises(BFRuntimeError):
            interpreter.readByte(0)

    def test_halted(self):
        interpreter = BFInterpreter(8, 16)
        interpreter.setMemory([], 0)
        interpreter.setTape(",+")

        assert interpreter.state == ProgramState.Ready
        assert not interpreter.halted()

        interpreter.step()

        assert interpreter.state == ProgramState.WaitingForInput
        assert not interpreter.halted()

        interpreter.readByte(5)
        assert interpreter.state == ProgramState.Running
        assert not interpreter.halted()

        interpreter.step()
        assert interpreter.state == ProgramState.Halted
        assert interpreter.halted()

        with pytest.raises(BFRuntimeError):
            interpreter.step()

        assert interpreter.state == ProgramState.Error
        assert interpreter.halted()

    def test_waitingForInput(self):
        interpreter = BFInterpreter(8, 16)
        interpreter.setMemory([], 0)
        interpreter.setTape(",+")

        assert interpreter.state == ProgramState.Ready
        assert not interpreter.waitingForInput()

        interpreter.step()

        assert interpreter.state == ProgramState.WaitingForInput
        assert interpreter.waitingForInput()

        interpreter.readByte(5)
        assert interpreter.state == ProgramState.Running
        assert not interpreter.waitingForInput()

        interpreter.step()
        assert interpreter.state == ProgramState.Halted
        assert not interpreter.waitingForInput()

        with pytest.raises(BFRuntimeError):
            interpreter.step()

        assert interpreter.state == ProgramState.Error
        assert not interpreter.waitingForInput()

    def test_canStep(self):
        interpreter = BFInterpreter(8, 16)
        interpreter.setMemory([], 0)
        interpreter.setTape(",+")

        assert interpreter.state == ProgramState.Ready
        assert interpreter.canStep()

        interpreter.step()

        assert interpreter.state == ProgramState.WaitingForInput
        assert not interpreter.canStep()

        interpreter.readByte(5)
        assert interpreter.state == ProgramState.Running
        assert interpreter.canStep()

        interpreter.step()
        assert interpreter.state == ProgramState.Halted
        assert not interpreter.canStep()

        with pytest.raises(BFRuntimeError):
            interpreter.step()

        assert interpreter.state == ProgramState.Error
        assert not interpreter.canStep()


tests/test_io_prompt.py
from src.io_prompt import IOPrompt
import pygame as pg


# Groups tests related to the IO Prompt
class TestIOPrompt:
    def initialize_pygame(self):
        if not pg.get_init():
            pg.init()

    def test_initialization_defaults(self):

        self.initialize_pygame()

        prompt = IOPrompt("test:")
        assert prompt.prompt == "test:"
        assert prompt.response == ""
        assert prompt.redraw is True
        assert prompt.toString() == "test: "

    def test_initialization_initresponse(self):

        self.initialize_pygame()

        prompt = IOPrompt("test:", "hello")
        assert prompt.prompt == "test:"
        assert prompt.response == "hello"
        assert prompt.redraw is True
        assert prompt.toString() == "test: hello"

    def test_setPrompt_empty(self):
        self.initialize_pygame()

        prompt = IOPrompt("test:")

        prompt.redraw = False
        prompt.setPrompt("")
        assert prompt.prompt == ""
        assert prompt.response == ""
        assert prompt.redraw is True

    def test_setPrompt_newValue(self):
        self.initialize_pygame()

        prompt = IOPrompt("test:")
        assert prompt.prompt == "test:"

        prompt.redraw = False
        prompt.setPrompt("newtest:")
        assert prompt.prompt == "newtest:"
        assert prompt.response == ""
        assert prompt.redraw is True
        assert prompt.toString() == "newtest: "

    def test_setResponse_empty(self):
        self.initialize_pygame()

        prompt = IOPrompt("test:", "hello")
        assert prompt.response == "hello"

        prompt.redraw = False
        prompt.setResponse("")
        assert prompt.prompt == "test:"
        assert prompt.response == ""
        assert prompt.redraw is True
        assert prompt.toString() == "test: "

    def test_setResponse_newValue(self):
        self.initialize_pygame()

        prompt = IOPrompt("test:")
        assert prompt.response == ""

        prompt.redraw = False
        prompt.setResponse("hello")
        assert prompt.prompt == "test:"
        assert prompt.response == "hello"
        assert prompt.redraw is True
        assert prompt.toString() == "test: hello"

    def test_appendResponse_empty(self):
        self.initialize_pygame()

        prompt = IOPrompt("test:", "hello")
        assert prompt.response == "hello"

        prompt.redraw = False
        prompt.appendResponse("")
        assert prompt.prompt == "test:"
        assert prompt.response == "hello"
        assert prompt.redraw is True
        assert prompt.toString() == "test: hello"

    def test_appendResponse_newText(self):
        self.initialize_pygame()

        prompt = IOPrompt("test:", "hello")
        assert prompt.response == "hello"

        prompt.redraw = False
        prompt.appendResponse(" world")
        assert prompt.prompt == "test:"
        assert prompt.response == "hello world"
        assert prompt.redraw is True
        assert prompt.toString() == "test: hello world"

    def test_backspace_empty(self):
        self.initialize_pygame()

        prompt = IOPrompt("test:")
        assert prompt.response == ""

        prompt.redraw = False
        prompt.backspaceResponse()
        assert prompt.prompt == "test:"
        assert prompt.response == ""
        assert prompt.redraw is False

    def test_backspace_withText(self):
        self.initialize_pygame()

        prompt = IOPrompt("test:", "hello")
        assert prompt.response == "hello"

        prompt.redraw = False
        prompt.backspaceResponse()
        assert prompt.prompt == "test:"
        assert prompt.response == "hell"
        assert prompt.redraw is True


src/gamestate.py
class Gamestate():

    maxHertz = 32
    minHertz = 1

    def __init__(self, step_hertz: int):

        self.step_hertz = step_hertz

    def setStepHertz(self, step_hertz: int, loop_on_overflow: bool = False, loop_on_underflow: bool = False):

        self.step_hertz = step_hertz

        if self.step_hertz < self.minHertz:
            if loop_on_underflow:
                self.step_hertz = self.maxHertz
            else:
                self.step_hertz = self.minHertz
        elif self.step_hertz > self.maxHertz:
            if loop_on_overflow:
                self.step_hertz = self.minHertz
            else:
                self.step_hertz = self.maxHertz

    def multiplyStepHertz(self, factor: float = 2.0, loop: bool = True):
        self.setStepHertz(int(self.step_hertz * factor), loop_on_overflow=loop, loop_on_underflow=loop)

    def incrementStepHertz(self, increment: int = 1, loop: bool = True):
        self.setStepHertz(self.step_hertz + increment, loop_on_overflow=loop, loop_on_underflow=loop)


src/io_prompt.py
import pygame as pg
import src.rendering_contants as rc


class IOPromptError(Exception):
    pass


class IOPrompt():

    def __init__(self, prompt: str, def_response: str = "", typeface: str = "freesansbold.ttf", point_size: int = 32):
        self.prompt = prompt
        self.response = def_response
        self.font = pg.font.Font(typeface, point_size)

        self.text_surface = pg.Surface = None
        self.redraw = True

    def renderPrompt(self, screen: pg.Surface, prompt_rect: pg.Rect):

        if self.redraw:
            self.text_surface = self.font.render(self.toString(), True, rc.CLR_WHITE, rc.CLR_BLACK)
            self.redraw = False

        pg.draw.rect(screen, rc.CLR_RED, prompt_rect, width=1)

        text_rect = self.text_surface.get_rect()
        text_rect.topleft = (prompt_rect.left + 5, prompt_rect.top + 5)
        screen.blit(self.text_surface, text_rect)

    def appendResponse(self, text: str):
        self.response = f"{self.response}{text}"
        self.redraw = True

    def setResponse(self, text: str):
        self.response = text
        self.redraw = True

    def setPrompt(self, prompt: str):
        self.prompt = prompt
        self.redraw = True

    def backspaceResponse(self):
        if len(self.response) > 0:
            self.response = self.response[:len(self.response)-1]
            self.redraw = True

    def toString(self):
        return f"{self.prompt} {self.response}"


src/hud_render.py

import pygame as pg
import src.rendering_contants as rc
from src.bf import BFInterpreter


class HudRenderer():

    def __init__(self, interpreter: BFInterpreter, typeface: str = "freesansbold.ttf", point_size: int = 32):

        self.interpreter = interpreter

        self.font = pg.font.Font(typeface, point_size)
        self.last_hertz = -1
        self.last_interpreter_state = -1
        self.last_step_count = -1

        self.hertz_text_surface: pg.Surface = None
        self.step_text_surface: pg.Surface = None
        self.state_text_surface: pg.Surface = None

    def renderHud(self, screen: pg.Surface, hud_rect: pg.Rect, hertz: int):
        if self.last_hertz != hertz:
            self.hertz_text_surface = self.font.render(f"CPU: {hertz}Hz", True, rc.CLR_WHITE, rc.CLR_BLACK)
            self.last_hertz = hertz

        if self.interpreter.state != self.last_interpreter_state:
            self.state_text_surface = self.font.render(f"State: {self.interpreter.state._name_}", True, rc.CLR_WHITE, rc.CLR_BLACK) # noqa 
            self.last_interpreter_state = self.interpreter.state

        if self.last_step_count != self.interpreter.step_count:
            self.step_text_surface = self.font.render(f"Step Count: {self.interpreter.step_count}", True, rc.CLR_WHITE, rc.CLR_BLACK) # noqa
            self.last_step_count = self.interpreter.step_count

        hertz_rect = self.hertz_text_surface.get_rect()
        state_rect = self.state_text_surface.get_rect()
        step_rect = self.step_text_surface.get_rect()

        hertz_rect.topleft = (10 + hud_rect.left, 10 + hud_rect.top)
        state_rect.topleft = (hertz_rect.left, hertz_rect.bottom + 10)
        step_rect.topleft = (state_rect.left, state_rect.bottom + 10)

        screen.blit(self.hertz_text_surface, hertz_rect)
        screen.blit(self.state_text_surface, state_rect)
        screen.blit(self.step_text_surface, step_rect)


main.py

import pygame as pg
import os
import sys
import logging
import time
import json
from src.bf import BFInterpreter, BFRuntimeError, ProgramState
from src.hud_render import HudRenderer
from src.interpreter_render import BFRenderer
from src.io_prompt import IOPrompt
import src.rendering_contants as rc
from src.gamestate import Gamestate


class CliInitError(Exception):
    pass


class EnvironmentInitError(Exception):
    pass


# see if we can load more than standard BMP
if not pg.image.get_extended():
    raise SystemExit("Sorry, extended image module required")

# Constants
# Rendering Display
SCREENRECT = pg.Rect(0, 0, 640, 480)

# UI Control Values
UI_INIT_TOPLEFT = (20, 20)

# Interpreter Control Values
INT_SRC = ""
INT_CELL_COUNT = 8
INT_CELL_MAX_VALUE = 16
INT_CELL_DEFAULT_VALUE = 0
INT_INITIAL_VALUES = []

main_dir = os.path.split(os.path.abspath(__file__))[0]


def load_image(file):

    file = os.path.join(main_dir, "images", file)
    try:
        surface = pg.image.load(file)
    except pg.error:
        raise SystemExit(f'Could not load image "{file}" {pg.get_error()}')
    return surface.convert()


def initPyGame():
    logging.info("Initializing PyGame")
    # Initialize pygame
    if pg.get_sdl_version()[0] == 2:
        pg.mixer.pre_init(44100, 32, 2, 1024)
    pg.init()
    if pg.mixer and not pg.mixer.get_init():
        logging.warning("Warning, no sound")
        pg.mixer = None


def initWindow(width=800, height=600) -> pg.Surface:
    global SCREENRECT

    SCREENRECT = pg.Rect(0, 0, width, height)
    winstyle = 0  # |FULLSCREEN
    bestdepth = pg.display.mode_ok(SCREENRECT.size, winstyle, 32)
    return pg.display.set_mode(SCREENRECT.size, winstyle, bestdepth)


def initLogging(verbose: bool):
    if verbose:
        logging.basicConfig(level=logging.DEBUG)
    else:
        logging.basicConfig(level=logging.INFO)


def openSrcFile(path: str):
    global INT_SRC
    logging.info(f"Loading source file from: {path}")

    with open(path, 'r') as f:
        INT_SRC = f.read()


def raiseEnvFileException(message: str):
    logging.error(message)
    raise EnvironmentInitError(message)


def openEnvFile(path: str):
    global INT_CELL_COUNT, INT_CELL_MAX_VALUE, INT_CELL_DEFAULT_VALUE, INT_INITIAL_VALUES

    logging.info(f"Loading environment file from: {path}")

    with open(path, 'r') as f:
        content_json = json.loads(f.read())

        # Setting cell maximum allowed value
        cell_max = content_json["memory"]["cell_max_value"]
        if type(cell_max) is not int or cell_max < 1:
            raiseEnvFileException("Value 'cell_max_value' must be an integer greater than 0")
        INT_CELL_MAX_VALUE = cell_max

        # Setting cell count
        cell_count = content_json["memory"]["cell_count"]
        if type(cell_count) is not int or cell_count < 1:
            raiseEnvFileException("Value 'cell_count' must be an integer greater than 0")
        INT_CELL_COUNT = cell_count

        # Setting cell default values
        cell_default = content_json["memory"]["cell_default_value"]
        if type(cell_default) is not int or cell_default < 0 or cell_default > INT_CELL_MAX_VALUE:
            raiseEnvFileException(f"Value 'cell_default_value' must be an integer in the range -> {INT_CELL_MAX_VALUE}")
        INT_CELL_DEFAULT_VALUE = cell_default

        # Setting cell initial values
        initial_values = content_json["memory"]["cell_initial_values"]

        # Do basic type and length validation with the settings already loaded
        if type(initial_values) is not list:
            raiseEnvFileException(f"List 'cell_initial_values' must be a list of integers in the range 0 -> {INT_CELL_MAX_VALUE}") # noqa

        if len(initial_values) > INT_CELL_COUNT:
            raiseEnvFileException("List 'cell_initial_values' cannot have more entries that the number of cells")

        for v in initial_values:
            if type(v) is not int or v < 0 or v > INT_CELL_MAX_VALUE:
                raiseEnvFileException("Values in list 'cell_initial_values' must be integers in the range 0 -> {INT_CELL_MAX_VALUE}") # noqa
        INT_INITIAL_VALUES = initial_values


def processCLI():

    verbose = False

    for i in range(1, len(sys.argv)):
        if sys.argv[i] in ["-v", "--verbose"]:
            verbose = True

    initLogging(verbose)

    # Load Command Line Arguments
    last_cmd = ""
    for i in range(1, len(sys.argv)):
        # Prepare to process next parameter set. May be one or two tokens
        if last_cmd == "":
            # Ignore verbose commands
            if sys.argv[i] in ["-v", "--verbose"]:
                continue

            # Explicitely allow only - parameters that are supported
            elif sys.argv[i] in ["-s", "--src-file", "-e", "--env-file"]:
                last_cmd = sys.argv[i]

            else:
                message = f"Unexpected cli parameter '{sys.argv[i]}' found"
                logging.error(message)
                raise CliInitError(message)

        # Process the second token in a two token parameter set.
        elif last_cmd in ["-s", "--src-file"]:
            openSrcFile(sys.argv[i])
            last_cmd = ""

        elif last_cmd in ["-e", "--env-file"]:
            openEnvFile(sys.argv[i])
            last_cmd = ""
        else:
            message = f"Unexpected cli parameter '{last_cmd}' found"
            logging.error(message)
            raise CliInitError(message)


def main(winstyle=0):

    processCLI()

    # Init the engine and display window
    initPyGame()
    screen = initWindow(800, 600)

    # Init a BF program
    bf_interpreter = BFInterpreter(INT_CELL_COUNT, INT_CELL_MAX_VALUE)
    bf_interpreter.setMemory(INT_INITIAL_VALUES, INT_CELL_DEFAULT_VALUE)
    bf_interpreter.setTape(INT_SRC)

    # Init Graphics Handlers
    bf_renderer = BFRenderer(interpreter=bf_interpreter,
                             cell_width=50,
                             cell_buffer=25,
                             camera_speed=100,
                             cell_max_height=300)

    hud_renderer = HudRenderer(interpreter=bf_interpreter)
    gs = Gamestate(step_hertz=1)

    clock = pg.time.Clock()

    # Set up the step execution values
    step_delay = 1.0/gs.step_hertz
    step_next_time = time.time() + step_delay
    step_run = False

    # Set up input handling
    readbyte_prompt_running = False

    # Set up contextual UI elements
    readbyte_prompt = IOPrompt("Cell Value:")

    last_tick = time.time()
    
    while True:
        # Hold Framerate to 60 fps
        clock.tick(60)
        sys_time= time.time()
        
        # Store the seconds elapsed since the last tick
        tick_time = sys_time - last_tick
        last_tick = sys_time

        # Handle Input
        for event in pg.event.get():
            if event.type == pg.QUIT:
                return
            if event.type == pg.KEYDOWN and event.key == pg.K_ESCAPE:
                return
            if readbyte_prompt_running:
                if event.type == pg.KEYUP:

                    if (event.unicode in "0123456789"):
                        readbyte_prompt.appendResponse(event.unicode)
                    elif event.key == pg.K_BACKSPACE:
                        readbyte_prompt.backspaceResponse()
                    elif event.key == pg.K_RETURN and len(readbyte_prompt.response) > 0:
                        readbyte_prompt_running = False
                        bf_interpreter.readByte(int(readbyte_prompt.response))
                        step_next_time = time.time() + step_delay

            else:
                # Pressing SPACE will toggle the program run mode
                if event.type == pg.KEYUP and event.key == pg.K_SPACE:
                    step_run = not step_run
                    step_next_time = time.time() + step_delay

                    if bf_interpreter.halted():
                        logging.debug("Disabling auto-run on halted program")
                        step_run = False

                    logging.debug(f"Setting program auto exec to {step_run}")

                # Pressing TAB will increase the execution rate by powers of 2
                if event.type == pg.KEYUP and event.key == pg.K_TAB:
                    gs.multiplyStepHertz(factor=2, loop=True)

                    step_delay = 1.0/gs.step_hertz
                    step_next_time = time.time() + step_delay

        # Execute Instructions
        if step_run and bf_interpreter.canStep() and time.time() >= step_next_time:
            try:
                bf_interpreter.step()
                step_next_time = step_next_time + step_delay

            except BFRuntimeError as runtime_error:
                logging.warning(f"Program execution failed due to runtime error:\r\n\t{runtime_error}")

        # UI Elements

        if bf_interpreter.state == ProgramState.WaitingForInput and not readbyte_prompt_running:
            readbyte_prompt.setResponse("")
            readbyte_prompt_running = True

        # Render Screen
        screen.fill(rc.CLR_BLACK)

        # Render UI Elements
        hud_renderer.renderHud(screen, pg.Rect(50, 50, 0, 0), gs.step_hertz)

        if readbyte_prompt_running:
            readbyte_prompt.renderPrompt(screen, pg.Rect(100, 200, 400, 50))

        bf_renderer.render(screen, pg.Rect(50, 200, 0, 0), tick_time)

        # Flip the display      
        pg.display.flip()


# call the "main" function if running this script
if __name__ == "__main__":
    main()
    pg.quit()
